### PR TITLE
Integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by Dachrono**
# 🎉 🎈 In this *Milestone* 🙋‍♂️ 🚩 ‼ :

**Hi there** 👋

- Follow the best GitHub flow. ✅
- Document in a professional way. ✅
- No linters errors. ✅
- Solve n+1 issue. ✅
- Install Capybara and tools to test. ✅
- Create the required test. ✅

# To the code reviewer 👨‍💻
I really thank 🙇‍♂️ you, dear code reviewer 👨‍🎨 for dedicating your time ⌚ to check the PR 🧾 .

I kindly ask code reviewer 🤵 that if there is any issue in this project, please do list 📃 them in a descriptive 💡 manner and give your best suggestions if needed.

If you think some significant issues need to be changed ♻️ please kindly contact 📞 me through Zoom or Slack, I am available on Microverse UTC-6 time zone.

Screenshot before solve n+1 problem

Captura desde 2023-10-04 14-11-51

Screenshot after solve n+1 problem

Captura desde 2023-10-04 14-16-42

### CODE REVIEW

File: spec/features/user_index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    expect(page).to have_css("img[src='#{@user3.photo}']")
+  end
+
+  it 'I can see the number of posts each user has written' do
+    visit users_path
+    expect(page).to have_content("Number of posts: #{@user1.posts_counter}")
+    expect(page).to have_content("Number of posts: #{@user2.posts_counter}")
+    expect(page).to have_content("Number of posts: #{@user3.posts_counter}")
+  end
+end
```
Comment (abdulhamiid) [inline spec/features/user_index_spec.rb]: Please add the test for clicking links as per the [requirements](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md) 🛠😊

```
When I click on a user, I am redirected to that user's show page.
```

Effective unit tests help catch defects early in the process, improve code quality, and facilitate refactoring 👍

File: spec/features/user_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+  it 'I can see the users first 3 posts.' do
+    visit user_path(@user1)
+    expect(page).to have_link("Post # #{@post1.id}", href: user_post_path(@user1.id, @post1.id))
+  end
+
+  it 'I can see a button that lets me view all of a users posts.' do
+    visit user_path(@user1)
+    expect(page).to have_link('See all posts', href: user_posts_path(@user1), class: 'button_footer')
+  end
```
Comment (abdulhamiid) [inline spec/features/user_show_spec.rb:39]: Please add the test for this:
```
When I click a user's post, it redirects me to that post's show page.
```

File: spec/features/post_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,57 @@
+require 'rails_helper'
+
+RSpec.describe 'Posts index page', type: :feature do
+  before :each do
+    @user1 = User.create(name: 'Andy Zam', photo: 'https://i.pravatar.cc/150')
```
Comment (abdulhamiid) [inline spec/features/post_index_spec.rb:5]: Please add test for:

- I can see a section for pagination if there are more posts than fit on the view.
- When I click on a post, it redirects me to that post's show page.

File: spec/features/comment_form_spec.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  scenario 'User adds a new comment to a post' do
+    visit user_post_path(@user1, @post1)
+
+    fill_in 'Text', with: 'This is a new comment for the post'
+
+    click_button 'Add Comment'
+
+    expect(page).to have_content('This is a new comment for the post')
+  end
+end
```
Comment (abdulhamiid) [inline spec/features/comment_form_spec.rb:19]: Good job testing the comment creation. 🛠🤓

**Review by abdulhamiid — CHANGES_REQUESTED** [review]
# Status: Required Changes :recycle:
Hi Team 👋, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there! :running_man: :runner: 
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>

## Highlights 🌟

- ✔️ Write integration tests with Capybara gem.
- ✔️ Secure app from n+1 problems.

## Required Changes ♻️
_Please kindly check the comments under the review._

## Optional suggestions 🤷‍♂️
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._


Cheers, and Happy coding!👏👏👏

## Important Notes 📝
- Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
- Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/features/user_index_spec.rb (hunk 1 above)
Comment (Dachrono) [reply spec/features/user_index_spec.rb]: Solved

File: spec/features/user_show_spec.rb (hunk 2 above)
Comment (Dachrono) [reply spec/features/user_show_spec.rb:39]: Solved

File: spec/features/post_index_spec.rb (hunk 3 above)
Comment (Dachrono) [reply spec/features/post_index_spec.rb:5]: Solved

File: spec/features/comment_form_spec.rb (hunk 4 above)
Comment (Dachrono) [reply spec/features/comment_form_spec.rb:19]: Thank you we appreciate it

**Review by Shedrack-Sunday — APPROVED** [review]
Hi @Dachrono 

Your project is complete! There is nothing else to say other than... it's time to merge it :shipit:
Congratulations! 🎉

image


Highlights
- [x] Good project organization ✔️
- [x] Nice use of Rails testing features ✔️  
- [x] Kudus on meeting the requirements suggested by the previous rviwer✔️


Optional suggestions
Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application.


Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Pull request merged**